bftclient/src/bft_client.cpp

```diff
@@ -0,0 +1,162 @@
+// this product except in compliance with the Apache 2.0 License.
+//
+// This product may include a number of subcomponents with separate copyright notices and license
+// terms. Your use of these subcomponents is subject to the terms and conditions of the
+// subcomponent's license, as noted in the LICENSE file.
+
+#include "bftclient/bft_client.h"
+#include "bftengine/ClientMsgs.hpp"
+#include "assertUtils.hpp"
+
+namespace bft::client {
+
+// This function creates a ClientRequestMsg or a ClientPreProcessRequestMsg depending upon config.
+//
+// Since both of these are just instancds of a `ClientRequestMsgHeader` followed by the message
+// data, we construct them here, rather than relying on the type constructors embedded into the
+// bftEngine impl. This allows us to not have to link with the bftengine library, and also allows us
+// to return the messages as vectors with proper RAII based memory management.
+Msg MakeClientMsg(const RequestConfig& config, Msg&& request, bool read_only, uint16_t client_id) {
+  uint8_t flags = read_only ? READ_ONLY_REQ : EMPTY_FLAGS_REQ;
+  if (config.pre_execute) {
+    flags |= PRE_PROCESS_REQ;
+  }
+
+  auto header_size = sizeof(bftEngine::ClientRequestMsgHeader);
+
+  Msg msg(header_size + request.size());
+  bftEngine::ClientRequestMsgHeader* header = reinterpret_cast<bftEngine::ClientRequestMsgHeader*>(msg.data());
+  header->msgType = config.pre_execute ? PRE_PROCESS_REQUEST_MSG_TYPE : REQUEST_MSG_TYPE;
+  header->spanContextSize = 0;  // TODO (AJS): FIXME
+  header->idOfClientProxy = client_id;
+  header->flags = flags;
+  header->reqSeqNum = config.sequence_number;
+  header->requestLength = msg.size();
+  header->timeoutMilli = config.timeout.count();
+  header->cid_length = config.correlation_id.size();
+
+  // Copy the correlation ID
+  std::memcpy(msg.data() + header_size, config.correlation_id.data(), config.correlation_id.size());
+
+  // Copy the request data
+  std::memcpy(msg.data() + header_size + config.correlation_id.size(), request.data(), request.size());
+
+  return msg;
+}
+
+Reply Client::Send(const WriteConfig& config, Msg&& request) {
+  Assert(!outstanding_request_.has_value());
+  auto match_config = WriteConfigToMatchConfig(config);
+  bool read_only = false;
+  return Send(match_config, config.request, std::move(request), read_only);
+}
+
+Reply Client::Send(const ReadConfig& config, Msg&& request) {
+  Assert(!outstanding_request_.has_value());
+  auto match_config = ReadConfigToMatchConfig(config);
+  bool read_only = true;
+  return Send(match_config, config.request, std::move(request), read_only);
+}
+
+Reply Client::Send(const MatchConfig& match_config,
+                   const RequestConfig& request_config,
+                   Msg&& request,
+                   bool read_only) {
+  outstanding_request_ = Matcher(match_config);
+  receiver_.activate(request_config.max_reply_size);
+  auto msg = MakeClientMsg(request_config, std::move(request), read_only, config_.id.val);
```

Comment:
Since the message is used only as a const reference, it would be better to use `const auto&` to avoid possible `move`s or copies.

Reply:
Wow, I did not know that this use of const auto& actually extends the lifetime of the returned temporary. In this case I"m not sure it matters, as we'd get return value optimization anyway. But to make you happy I changed it. Thanks for pointing it out. 

For other reviewers see the section "Extending the lifetime of temporaries" here: https://crascit.com/2016/08/15/avoiding-copies-and-moves-with-auto/